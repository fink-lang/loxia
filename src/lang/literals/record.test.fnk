{fink2js} = import '../../testing/generate.fnk'
{describe, it, expect, to_match_snapshot} = import '@fink/jest/test.fnk'


describe 'record', fn:
  it 'compiles simple', fn:
    expect
      fink2js 'obj = {}'
      to_match_snapshot


  it 'compiles shorthand', fn:
    expect
      fink2js 'obj = {foo, π, ƒ}'
      to_match_snapshot

    expect
      fink2js 'obj = {a, delete, true, false}'
      to_match_snapshot

    expect
      fink2js "obj = {'foo'}"
      to_match_snapshot


  it 'compiles spread', fn:
    expect
      fink2js 'obj = {a, b, ...c}'
      to_match_snapshot


  it 'compiles key:val', fn:
    expect
      fink2js "obj = {a: 1, b: 123, 'c-d-e': cde}"
      to_match_snapshot

    expect
      fink2js '
        obj7 = {a: 123, b: 123 and 123 and 1345, c: fn a, b: 134}
      '
      to_match_snapshot

  it 'compiles idents that are not js prop-names', fn:
    expect
      fink2js "
        foo = rec:
          foo: bar
          π: 3.12,
          ƒ: 'fink'
        "
      to_match_snapshot

  it 'compiles multiline', fn:
    expect
      fink2js 'obj = {
          a: 123

          b: 123
            and 123
            and 1345

          c: fn a, b: 134

          d: foo a + 3
        }
      '
      to_match_snapshot

    expect
      fink2js 'obj = rec:
          a: 123

          b: 123
            and 123
            and 1345

          c: fn a, b: 134

          d: foo a + 3
       '
      to_match_snapshot


describe 'calculated props', fn:
  it 'compiles', fn:
    expect
      fink2js "
        foo = rec:
          (Symbol 'foo'): 123
          'spam': 456
          'spam-\${ham}': 'ni'
        "
      to_match_snapshot




describe 'destructuring record', fn:
  it 'compiles simple', fn:
    expect
      fink2js '{a, b, c} = ni'
      to_match_snapshot


  it 'compiles non js idents', fn:
    expect
      fink2js '{foo, π, ƒ, delete} = obj'
      to_match_snapshot


  it 'compiles spread', fn:
    expect
      fink2js '{e, ...f} = {e: 0, foo: 12, bar: 34}'
      to_match_snapshot


  it 'compiles empty', fn:
    expect
      fink2js '{foo: _ , bar} = {foo: 12, bar: 34}'
      to_match_snapshot


  it 'compiles defaults', fn:
    expect
      fink2js '{x=1, π=2, ni: {y, z}} = {ni: {y: 1, z: 2}}'
      to_match_snapshot


  it 'compiles computed keys', fn:
    expect
      fink2js "{'n i': ni} = foo"
      to_match_snapshot


  # it 'compiles dot', fn:
  #   expect
  #     fink2js "foo = {...foo, a.b.c: 123, c.d.e: 13}"
  #     to_match_snapshot

  #   expect
  #     fink2js "{a.b.c, c.d.e, ...foo} = foo"
  #     to_match_snapshot


  # it 'compiles sub recs', fn:
  #   expect
  #     fink2js "{a: {...{a, b, c}: c, ...rest}} = foo"
  #     to_match_snapshot

  #   expect
  #     fink2js "foo = {a: {...{a, b, c}: c, ...rest}}"
  #     to_match_snapshot

