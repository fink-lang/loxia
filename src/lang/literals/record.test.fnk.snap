// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculated props compiles 1`] = `
"export const foo = {
  [\`spam\`]: 456,
  [\`spam-\${ham}\`]: \`ni\`
};"
`;

exports[`record compiles idents that are not js prop-names 1`] = `
"export const foo = {
  foo: bar,
  π: 3.12,
  ƒ: \`fink\`,
  \\"foo-bar\\": spam
};"
`;

exports[`record compiles key:val 1`] = `
"export const foo = {
  a: 1,
  b: 123,
  [\`c-d-e\`]: cde
};"
`;

exports[`record compiles key:val 2`] = `
"export const obj7 = {
  a: 123,
  b: 123 && 123 && 1345,
  c: (a, b) => 134
};"
`;

exports[`record compiles member expr as keys 1`] = `
"let _do_result;

{
  let ˆrec_1 = { ...bar
  };
  ˆrec_1 = { ...ˆrec_1,
    spam: { ...ˆrec_1?.spam,
      ham: { ...ˆrec_1?.spam?.ham,
        ni: \`ni\`
      }
    }
  };
  ˆrec_1 = { ...ˆrec_1,
    na: 1234
  };
  ˆrec_1 = { ...ˆrec_1,
    ...nu
  };
  _do_result = ˆrec_1;
}
export const foo = _do_result;
_do_result = undefined;"
`;

exports[`record compiles member expr as keys 2`] = `
"let _do_result;

{
  let ˆrec_1 = { ...bar
  };
  ˆrec_1 = { ...ˆrec_1,
    spam: { ...ˆrec_1?.spam,
      ham: (() => {
        let ˆrec_2 = {};
        ˆrec_2 = { ...ˆrec_2,
          ni: { ...ˆrec_2?.ni,
            na: \`ni\`
          }
        };
        return ˆrec_2;
      })()
    }
  };
  ˆrec_1 = { ...ˆrec_1,
    ...nu
  };
  _do_result = ˆrec_1;
}
export const foo = _do_result;
_do_result = undefined;"
`;

exports[`record compiles multiline 1`] = `
"export const foo = {
  a: 123,
  b: 123 && 123 && 1345,
  c: (a, b) => 134,
  d: foo(a + 3)
};"
`;

exports[`record compiles multiline 2`] = `
"export const foo = {
  a: 123,
  b: 123 && 123 && 1345,
  c: (a, b) => 134,
  d: foo(a + 3)
};"
`;

exports[`record compiles shorthand 1`] = `
"export const foo = {
  foo,
  π,
  ƒ,
  \\"foo-bar\\": fooᜭbar
};"
`;

exports[`record compiles shorthand 2`] = `
"export const foo = {
  a,
  \\"delete\\": ˆdelete,
  true,
  false
};"
`;

exports[`record compiles shorthand 3`] = `
"export const foo = {
  [\`foo\`]: \`foo\`
};"
`;

exports[`record compiles simple 1`] = `"export const foo = {};"`;

exports[`record compiles spread 1`] = `
"export const foo = {
  a,
  b,
  ...c
};"
`;

exports[`unpacking record compiles computed keys 1`] = `
"const {
  [\`n i\`]: ni
} = foo;"
`;

exports[`unpacking record compiles defaults 1`] = `
"const {
  x = 1,
  π = 2,
  ni: {
    y,
    z
  }
} = {
  ni: {
    y: 1,
    z: 2
  }
};"
`;

exports[`unpacking record compiles empty 1`] = `
"const {
  foo: ˆunused_1,
  bar
} = {
  foo: 12,
  bar: 34
};"
`;

exports[`unpacking record compiles member expr as keys 1`] = `
"const {
  spam: {
    ham: {
      ni: ni
    }
  },
  ...rest
} = foo;"
`;

exports[`unpacking record compiles member expr as keys 2`] = `
"const {
  spam: {
    ham: {
      ni: {
        ni: nu,
        foo: {
          bar: bar
        }
      }
    }
  },
  ...rest
} = foo;"
`;

exports[`unpacking record compiles member expr as keys 3`] = `
"const {
  \\"foo-bar\\": {
    \\"spam-ham\\": {
      ni: {
        ni: nu,
        foo: {
          bar: bar
        }
      }
    }
  },
  ...rest
} = foo;"
`;

exports[`unpacking record compiles non js idents 1`] = `
"const {
  foo,
  π,
  ƒ,
  \\"delete\\": ˆdelete
} = spam;"
`;

exports[`unpacking record compiles simple 1`] = `
"const {
  a,
  b,
  c
} = ni;"
`;

exports[`unpacking record compiles spread 1`] = `
"const {
  e,
  ...f
} = {
  e: 0,
  foo: 12,
  bar: 34
};"
`;
